src/rebar_state.erl

```diff
@@ -279,15 +278,43 @@ add_provider(State=#state_t{providers=Providers}, Provider) ->
     State#state_t{providers=[Provider | Providers]}.
 
 create_logic_providers(ProviderModules, State0) ->
-    lists:foldl(fun(ProviderMod, Acc) ->
-                        case providers:new(ProviderMod, Acc) of
-                            {error, Reason} ->
-                                ?ERROR(Reason++"~n", []),
-                                Acc;
-                            {ok, State1} ->
-                                State1
-                        end
-                end, State0, ProviderModules).
+    State1 = try
+                 lists:foldl(fun(ProviderMod, StateAcc) ->
+                                     case providers:new(ProviderMod, StateAcc) of
+                                         {error, Reason} ->
+                                             ?ERROR(Reason++"~n", []),
+                                             StateAcc;
+                                         {ok, StateAcc1} ->
+                                             StateAcc1
+                                     end
+                             end, State0, ProviderModules)
+             catch
+                 C:T ->
+                     ?DEBUG("~p: ~p ~p", [C, T, erlang:get_stacktrace()]),
+                     throw({error, "Failed creating providers. Run with DEBUG=1 for stacktrace."})
+             end,
+    apply_hooks(State1).
```

Comment:
Can you rewrite this block to be like:

```erlang
create_logic_providers(ProviderModules, State0) ->
    try
        State = lists:foldl(fun(ProviderMod, StateAcc) ->
                            case providers:new(ProviderMod, StateAcc) of
                                {error, Reason} ->
                                    ?ERROR(Reason++"~n", []),
                                    StateAcc;
                                {ok, StateAcc1} ->
                                    StateAcc1
                            end
                    end, State0, ProviderModules)
        apply_hooks(State)
    catch
        C:T ->
            ?DEBUG("~p: ~p ~p", [C, T, erlang:get_stacktrace()]),
            throw({error, "Failed creating providers. Run with DEBUG=1 for stacktrace."})
    end.
```

Or something like that to limit how rapidly the indenting grows in files and make the non-execution of hooks depend on the exception rather than the rethrow?

I'm also wondering if we shouldn't stop the fold whenever we hit an error with one of the providers rather than just pushing through. Is this something we should consider doing?